File: app/services/social_service.py
"""Social service for user social interactions."""

from uuid import UUID

from fastapi import HTTPException, status
from sqlalchemy import desc, func, select
from sqlalchemy.exc import DisconnectionError
from sqlalchemy.exc import TimeoutError as SQLTimeoutError

from app.api.v1.schemas.common.user import User as UserSchema
from app.api.v1.schemas.response.social import FollowResponse, GetFollowedUsersResponse
from app.api.v1.schemas.response.user_activity.favorite_summary import FavoriteSummary
from app.api.v1.schemas.response.user_activity.recipe_summary import RecipeSummary
from app.api.v1.schemas.response.user_activity.review_summary import ReviewSummary
from app.api.v1.schemas.response.user_activity.user_activity_response import (
    UserActivityResponse,
)
from app.api.v1.schemas.response.user_activity.user_summary import UserSummary
from app.core.logging import get_logger
from app.db.sql.models.recipe.recipe import Recipe
from app.db.sql.models.recipe.recipe_favorite import RecipeFavorite
from app.db.sql.models.recipe.recipe_review import RecipeReview
from app.db.sql.models.user.user import User
from app.db.sql.models.user.user_follows import UserFollows
from app.db.sql.sql_database_session import SqlDatabaseSession
from app.exceptions.custom_exceptions.database_exceptions import DatabaseError
from app.utils.privacy import PrivacyChecker

_log = get_logger(__name__)


class SocialService:
    """Service for social operations."""

    def __init__(self, db: SqlDatabaseSession) -> None:
        """Initialize social service with database session."""
        self.db = db
        self.privacy_checker = PrivacyChecker(db)

    async def get_followed_users(
        self,
        user_id: UUID,
        requester_user_id: UUID,
        limit: int = 20,
        offset: int = 0,
        count_only: bool = False,
    ) -> GetFollowedUsersResponse:
        """Get users that a user is following, respecting privacy preferences."""
        _log.info(f"Getting following list for user: {user_id}")
        try:
            user = await self.db.get_user_by_id(str(user_id))
            if not user:
                _log.warning(f"User not found: {user_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="User not found",
                )
            if (
                requester_user_id is not None
                and not await self.privacy_checker.check_access(user, requester_user_id)
            ):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Access denied due to privacy settings",
                )

            # Get total count (before filtering)
            count_result = await self.db.execute(
                select(func.count(UserFollows.followee_id)).where(
                    UserFollows.follower_id == user_id
                )
            )
            total_count = count_result.scalar()
            _log.info(f"Following count for user {user_id}: {total_count}")

            if count_only:
                return GetFollowedUsersResponse(
                    total_count=total_count,
                    followed_users=None,
                    limit=None,
                    offset=None,
                )

            # Get following relationships with user details
            following_result = await self.db.execute(
                select(User)
                .join(UserFollows, User.user_id == UserFollows.followee_id)
                .where(UserFollows.follower_id == user_id)
                .order_by(UserFollows.followed_at.desc())
                .limit(limit)
                .offset(offset)
            )
            following_users = following_result.scalars().all()

            # Per-user privacy filtering
            filtered_users = [
                u
                for u in following_users
                if await self.privacy_checker.check_access(u, requester_user_id)
            ]
            user_schemas = [
                UserSchema.model_validate(u).model_copy(update={"email": None})
                for u in filtered_users
            ]
            _log.info(
                f"Retrieved {len(user_schemas)} following users for user {user_id} "
                f"(after privacy filtering)"
            )

            return GetFollowedUsersResponse(
                total_count=total_count,
                followed_users=user_schemas,
                limit=limit,
                offset=offset,
            )

        except DatabaseError as e:
            _log.error(f"Database error while getting following list: {e}")
            raise HTTPException(
                status_code=e.status_code,
                detail=str(e),
            ) from e
        except (DisconnectionError, SQLTimeoutError) as e:
            _log.error(f"Database connection error while getting following list: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "Database service is temporarily unavailable. "
                    "Please try again later."
                ),
            ) from e

    async def get_followers(
        self,
        user_id: UUID,
        requester_user_id: UUID,
        limit: int = 20,
        offset: int = 0,
        count_only: bool = False,
    ) -> GetFollowedUsersResponse:
        """Get users who follow the given user, respecting privacy preferences."""
        _log.info(f"Getting followers list for user: {user_id}")
        try:
            user = await self.db.get_user_by_id(str(user_id))
            if not user:
                _log.warning(f"User not found: {user_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="User not found",
                )
            if (
                requester_user_id is not None
                and not await self.privacy_checker.check_access(user, requester_user_id)
            ):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Access denied due to privacy settings",
                )

            # Get total count (before filtering)
            count_result = await self.db.execute(
                select(func.count(UserFollows.follower_id)).where(
                    UserFollows.followee_id == user_id
                )
            )
            total_count = count_result.scalar()
            _log.info(f"Followers count for user {user_id}: {total_count}")

            if count_only:
                return GetFollowedUsersResponse(
                    total_count=total_count,
                    followed_users=None,
                    limit=None,
                    offset=None,
                )

            # Get follower relationships with user details
            followers_result = await self.db.execute(
                select(User)
                .join(UserFollows, User.user_id == UserFollows.follower_id)
                .where(UserFollows.followee_id == user_id)
                .order_by(UserFollows.followed_at.desc())
                .limit(limit)
                .offset(offset)
            )
            followers = followers_result.scalars().all()

            # Per-user privacy filtering
            filtered_users = [
                u
                for u in followers
                if await self.privacy_checker.check_access(u, requester_user_id)
            ]
            user_schemas = [
                UserSchema.model_validate(u).model_copy(update={"email": None})
                for u in filtered_users
            ]
            _log.info(
                f"Retrieved {len(user_schemas)} followers for user {user_id} "
                f"(after privacy filtering)"
            )

            return GetFollowedUsersResponse(
                total_count=total_count,
                followed_users=user_schemas,
                limit=limit,
                offset=offset,
            )

        except DatabaseError as e:
            _log.error(f"Database error while getting followers list: {e}")
            raise HTTPException(
                status_code=e.status_code,
                detail=str(e),
            ) from e
        except (DisconnectionError, SQLTimeoutError) as e:
            _log.error(f"Database connection error while getting followers list: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "Database service is temporarily unavailable. "
                    "Please try again later."
                ),
            ) from e

    async def follow_user(
        self, follower_id: UUID, target_user_id: UUID
    ) -> FollowResponse:
        """Follow a user.

        Args:
            follower_id: The user who wants to follow
            target_user_id: The user to be followed

        Returns:
            FollowResponse: Confirmation of following action

        Raises:
            HTTPException: If users not found, already following, or database error
        """
        _log.info(f"User {follower_id} attempting to follow user {target_user_id}")

        try:
            # Check if both users exist
            follower = await self.db.get_user_by_id(str(follower_id))
            if not follower:
                _log.warning(f"Follower not found: {follower_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Follower not found",
                )

            target_user = await self.db.get_user_by_id(str(target_user_id))
            if not target_user:
                _log.warning(f"Target user not found: {target_user_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Target user not found",
                )

            # Check if users are active
            if not follower.is_active:
                _log.warning(f"Inactive follower: {follower_id}")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Follower account is inactive",
                )

            if not target_user.is_active:
                _log.warning(f"Inactive target user: {target_user_id}")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Target user account is inactive",
                )

            # Prevent self-following
            if follower_id == target_user_id:
                _log.warning(f"User {follower_id} attempted to follow themselves")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Cannot follow yourself",
                )

            # Check if already following
            existing_follow = await self.db.execute(
                select(UserFollows).where(
                    UserFollows.follower_id == follower_id,
                    UserFollows.followee_id == target_user_id,
                )
            )
            if existing_follow.scalar_one_or_none():
                _log.info(f"User {follower_id} already following {target_user_id}")
                return FollowResponse(
                    message="Already following this user",
                    is_following=True,
                )

            # Create follow relationship
            follow_relationship = UserFollows(
                follower_id=follower_id,
                followee_id=target_user_id,
            )
            self.db.add(follow_relationship)
            await self.db.commit()

            _log.info(f"User {follower_id} successfully followed {target_user_id}")
            return FollowResponse(
                message="Successfully followed user",
                is_following=True,
            )

        except DatabaseError as e:
            _log.error(f"Database error while following user: {e}")
            raise HTTPException(
                status_code=e.status_code,
                detail=str(e),
            ) from e
        except (DisconnectionError, SQLTimeoutError) as e:
            _log.error(f"Database connection error while following user: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "Database service is temporarily unavailable. "
                    "Please try again later."
                ),
            ) from e

    async def unfollow_user(
        self, follower_id: UUID, target_user_id: UUID
    ) -> FollowResponse:
        """Unfollow a user.

        Args:
            follower_id: The user who wants to unfollow
            target_user_id: The user to be unfollowed

        Returns:
            FollowResponse: Confirmation of unfollowing action

        Raises:
            HTTPException: If users not found, not following, or database error
        """
        _log.info(f"User {follower_id} attempting to unfollow user {target_user_id}")

        try:
            # Check if both users exist
            follower = await self.db.get_user_by_id(str(follower_id))
            if not follower:
                _log.warning(f"Follower not found: {follower_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Follower not found",
                )

            target_user = await self.db.get_user_by_id(str(target_user_id))
            if not target_user:
                _log.warning(f"Target user not found: {target_user_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Target user not found",
                )

            # Check if users are active
            if not follower.is_active:
                _log.warning(f"Inactive follower: {follower_id}")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Follower account is inactive",
                )

            if not target_user.is_active:
                _log.warning(f"Inactive target user: {target_user_id}")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Target user account is inactive",
                )

            # Prevent self-unfollowing
            if follower_id == target_user_id:
                _log.warning(f"User {follower_id} attempted to unfollow themselves")
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Cannot unfollow yourself",
                )

            # Check if currently following
            existing_follow = await self.db.execute(
                select(UserFollows).where(
                    UserFollows.follower_id == follower_id,
                    UserFollows.followee_id == target_user_id,
                )
            )
            follow_relationship = existing_follow.scalar_one_or_none()
            if not follow_relationship:
                _log.info(f"User {follower_id} not following {target_user_id}")
                return FollowResponse(
                    message="Not following this user",
                    is_following=False,
                )

            # Remove follow relationship
            await self.db.delete(follow_relationship)
            await self.db.commit()

            _log.info(f"User {follower_id} successfully unfollowed {target_user_id}")
            return FollowResponse(
                message="Successfully unfollowed user",
                is_following=False,
            )

        except DatabaseError as e:
            _log.error(f"Database error while unfollowing user: {e}")
            raise HTTPException(
                status_code=e.status_code,
                detail=str(e),
            ) from e
        except (DisconnectionError, SQLTimeoutError) as e:
            _log.error(f"Database connection error while unfollowing user: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "Database service is temporarily unavailable. "
                    "Please try again later."
                ),
            ) from e

    async def get_user_activity(
        self,
        user_id: UUID,
        requester_user_id: UUID,
        per_type_limit: int = 15,
    ) -> UserActivityResponse:
        """Get recent user activity (recipes, follows, reviews, favorites).

        Args:
            user_id: The user's unique identifier
            requester_user_id: The user making the request
            per_type_limit: Number of results per activity type

        Returns:
            UserActivityResponse: Aggregated user activity

        Raises:
            HTTPException: If user not found, forbidden, or DB error
        """
        _log.info(f"Getting activity for user: {user_id}")
        try:
            user = await self.db.get_user_by_id(str(user_id))
            if not user:
                _log.warning(f"User not found: {user_id}")
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="User not found",
                )
            if not await self.privacy_checker.check_access(user, requester_user_id):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Access denied due to privacy settings",
                )

            # Recipes
            recipes_q = (
                select(Recipe)
                .where(Recipe.user_id == user_id)
                .order_by(desc(Recipe.created_at))
                .limit(per_type_limit)
            )
            recipes_result = await self.db.execute(recipes_q)
            recipes = recipes_result.scalars().all()

            # Follows (recently followed users)
            follows_q = (
                select(UserFollows)
                .where(UserFollows.follower_id == user_id)
                .order_by(desc(UserFollows.followed_at))
                .limit(per_type_limit)
            )
            follows_result = await self.db.execute(follows_q)
            follows = follows_result.scalars().all()

            # Reviews
            reviews_q = (
                select(RecipeReview)
                .where(RecipeReview.user_id == user_id)
                .order_by(desc(RecipeReview.created_at))
                .limit(per_type_limit)
            )
            reviews_result = await self.db.execute(reviews_q)
            reviews = reviews_result.scalars().all()

            # Favorites
            favorites_q = (
                select(RecipeFavorite)
                .where(RecipeFavorite.user_id == user_id)
                .order_by(desc(RecipeFavorite.favorited_at))
                .limit(per_type_limit)
            )
            favorites_result = await self.db.execute(favorites_q)
            favorites = favorites_result.scalars().all()

            # Transform to response schemas
            recipe_summaries = [
                RecipeSummary.model_validate(recipe) for recipe in recipes
            ]
            review_summaries = [
                ReviewSummary.model_validate(review) for review in reviews
            ]
            favorite_summaries = [
                FavoriteSummary.model_validate(favorite) for favorite in favorites
            ]

            # For follows, we need to get the followed user details
            followed_users = []
            for follow in follows:
                followed_user = await self.db.get_user_by_id(str(follow.followee_id))
                if followed_user:
                    followed_users.append(
                        UserSummary.model_validate(follow).model_copy(
                            update={
                                "user_id": followed_user.user_id,
                                "username": followed_user.username,
                            }
                        )
                    )

            return UserActivityResponse(
                user_id=user_id,
                recent_recipes=recipe_summaries,
                recent_follows=followed_users,
                recent_reviews=review_summaries,
                recent_favorites=favorite_summaries,
            )

        except DatabaseError as e:
            _log.error(f"Database error while getting user activity: {e}")
            raise HTTPException(
                status_code=e.status_code,
                detail=str(e),
            ) from e
        except (DisconnectionError, SQLTimeoutError) as e:
            _log.error(f"Database connection error while getting user activity: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail=(
                    "Database service is temporarily unavailable. "
                    "Please try again later."
                ),
            ) from e

File: internal/service/social_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jsamuelsen/recipe-web-app/user-management-service/internal/dto"
	"github.com/jsamuelsen/recipe-web-app/user-management-service/internal/notification"
	"github.com/jsamuelsen/recipe-web-app/user-management-service/internal/repository"
)

// SocialService defines business logic for social operations.
type SocialService interface {
	GetFollowing(
		ctx context.Context,
		requesterID, targetUserID uuid.UUID,
		limit, offset int,
		countOnly bool,
	) (*dto.GetFollowedUsersResponse, error)
	GetFollowers(
		ctx context.Context,
		requesterID, targetUserID uuid.UUID,
		limit, offset int,
		countOnly bool,
	) (*dto.GetFollowedUsersResponse, error)
	FollowUser(
		ctx context.Context,
		followerID, targetUserID uuid.UUID,
	) (*dto.FollowResponse, error)
	UnfollowUser(
		ctx context.Context,
		followerID, targetUserID uuid.UUID,
	) (*dto.FollowResponse, error)
	GetUserActivity(
		ctx context.Context,
		requesterID *uuid.UUID,
		targetUserID uuid.UUID,
		perTypeLimit int,
	) (*dto.UserActivityResponse, error)
}

// ErrAccessDenied is returned when access to a resource is denied due to privacy settings.
var ErrAccessDenied = errors.New("access denied")

// ErrCannotFollowSelf is returned when a user tries to follow themselves.
var ErrCannotFollowSelf = errors.New("cannot follow yourself")

// ErrFollowNotAllowed is returned when target user has disabled follows.
var ErrFollowNotAllowed = errors.New("user does not allow follows")

// ErrCannotUnfollowSelf is returned when a user tries to unfollow themselves.
var ErrCannotUnfollowSelf = errors.New("cannot unfollow yourself")

// Profile visibility constants.
const (
	profileVisibilityPublic        = "public"
	profileVisibilityFollowersOnly = "followers_only"
	profileVisibilityPrivate       = "private"
)

// SocialServiceImpl implements SocialService.
type SocialServiceImpl struct {
	userRepo           repository.UserRepository
	socialRepo         repository.SocialRepository
	notificationClient notification.Client
}

// NewSocialService creates a new SocialService.
func NewSocialService(
	userRepo repository.UserRepository,
	socialRepo repository.SocialRepository,
	notificationClient notification.Client,
) *SocialServiceImpl {
	return &SocialServiceImpl{
		userRepo:           userRepo,
		socialRepo:         socialRepo,
		notificationClient: notificationClient,
	}
}

// GetFollowing retrieves the list of users that the target user follows.
func (s *SocialServiceImpl) GetFollowing(
	ctx context.Context,
	requesterID, targetUserID uuid.UUID,
	limit, offset int,
	countOnly bool,
) (*dto.GetFollowedUsersResponse, error) {
	// 1. Verify target user exists
	user, err := s.userRepo.FindUserByID(ctx, targetUserID)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrUserNotFound
		}

		return nil, fmt.Errorf("failed to fetch user: %w", err)
	}

	// 2. Check if user is active
	if !user.IsActive {
		return nil, ErrUserNotFound
	}

	// 3. Check privacy settings
	canAccess, err := s.canAccessFollowingList(ctx, requesterID, targetUserID)
	if err != nil {
		return nil, err
	}

	if !canAccess {
		return nil, ErrAccessDenied
	}

	// 4. Get following list from repository
	users, totalCount, err := s.socialRepo.GetFollowing(ctx, targetUserID, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("failed to get following list: %w", err)
	}

	// 5. Build response
	return s.buildFollowingResponse(users, totalCount, limit, offset, countOnly), nil
}

// GetFollowers retrieves the list of users who follow the target user.
func (s *SocialServiceImpl) GetFollowers(
	ctx context.Context,
	requesterID, targetUserID uuid.UUID,
	limit, offset int,
	countOnly bool,
) (*dto.GetFollowedUsersResponse, error) {
	// 1. Verify target user exists
	user, err := s.userRepo.FindUserByID(ctx, targetUserID)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrUserNotFound
		}

		return nil, fmt.Errorf("failed to fetch user: %w", err)
	}

	// 2. Check if user is active
	if !user.IsActive {
		return nil, ErrUserNotFound
	}

	// 3. Check privacy settings (same rules as following list)
	canAccess, err := s.canAccessFollowingList(ctx, requesterID, targetUserID)
	if err != nil {
		return nil, err
	}

	if !canAccess {
		return nil, ErrAccessDenied
	}

	// 4. Get followers list from repository
	users, totalCount, err := s.socialRepo.GetFollowers(ctx, targetUserID, limit, offset)
	if err != nil {
		return nil, fmt.Errorf("failed to get followers list: %w", err)
	}

	// 5. Build response
	return s.buildFollowingResponse(users, totalCount, limit, offset, countOnly), nil
}

// FollowUser creates a follow relationship from follower to target user.
func (s *SocialServiceImpl) FollowUser(
	ctx context.Context,
	followerID, targetUserID uuid.UUID,
) (*dto.FollowResponse, error) {
	// 1. Check self-follow
	if followerID == targetUserID {
		return nil, ErrCannotFollowSelf
	}

	// 2. Verify target user exists and is active
	targetUser, err := s.userRepo.FindUserByID(ctx, targetUserID)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrUserNotFound
		}

		return nil, fmt.Errorf("failed to fetch target user: %w", err)
	}

	if !targetUser.IsActive {
		return nil, ErrUserNotFound
	}

	// 3. Check privacy settings - if AllowFollows is false, return forbidden
	privacy, err := s.userRepo.FindPrivacyPreferencesByUserID(ctx, targetUserID)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch privacy preferences: %w", err)
	}

	if !privacy.AllowFollows {
		return nil, ErrFollowNotAllowed
	}

	// 4. Create follow relationship (idempotent - duplicate follows are OK)
	err = s.socialRepo.FollowUser(ctx, followerID, targetUserID)
	if err != nil {
		return nil, fmt.Errorf("failed to follow user: %w", err)
	}

	// 5. Send notification (fire-and-forget)
	// Use context.Background() to decouple from request context so notification
	// continues even if the request is cancelled.
	if s.notificationClient != nil {
		go s.notificationClient.NotifyNewFollower(context.Background(), targetUserID, followerID) //nolint:contextcheck
	}

	// 6. Return success response
	return &dto.FollowResponse{
		Message:     "Successfully followed user",
		IsFollowing: true,
	}, nil
}

// UnfollowUser removes a follow relationship from follower to target user.
func (s *SocialServiceImpl) UnfollowUser(
	ctx context.Context,
	followerID, targetUserID uuid.UUID,
) (*dto.FollowResponse, error) {
	// 1. Check self-unfollow
	if followerID == targetUserID {
		return nil, ErrCannotUnfollowSelf
	}

	// 2. Verify target user exists and is active
	targetUser, err := s.userRepo.FindUserByID(ctx, targetUserID)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrUserNotFound
		}

		return nil, fmt.Errorf("failed to fetch target user: %w", err)
	}

	if !targetUser.IsActive {
		return nil, ErrUserNotFound
	}

	// 3. Delete follow relationship (idempotent - success even if not following)
	err = s.socialRepo.UnfollowUser(ctx, followerID, targetUserID)
	if err != nil {
		return nil, fmt.Errorf("failed to unfollow user: %w", err)
	}

	// 4. Return success response
	return &dto.FollowResponse{
		Message:     "Successfully unfollowed user",
		IsFollowing: false,
	}, nil
}

// GetUserActivity retrieves a user's recent activity with privacy checks.
//
//nolint:cyclop,funlen // Complexity and length are from sequential error checks which are clear and necessary
func (s *SocialServiceImpl) GetUserActivity(
	ctx context.Context,
	requesterID *uuid.UUID,
	targetUserID uuid.UUID,
	perTypeLimit int,
) (*dto.UserActivityResponse, error) {
	// 1. Verify target user exists and is active
	user, err := s.userRepo.FindUserByID(ctx, targetUserID)
	if err != nil {
		if errors.Is(err, repository.ErrUserNotFound) {
			return nil, ErrUserNotFound
		}

		return nil, fmt.Errorf("failed to fetch user: %w", err)
	}

	if !user.IsActive {
		return nil, ErrUserNotFound
	}

	// 2. Check privacy settings
	canAccess, err := s.canAccessUserActivity(ctx, requesterID, targetUserID)
	if err != nil {
		return nil, err
	}

	if !canAccess {
		return nil, ErrAccessDenied
	}

	// 3. Fetch all activity data
	recipes, err := s.socialRepo.GetRecentRecipes(ctx, targetUserID, perTypeLimit)
	if err != nil {
		return nil, fmt.Errorf("failed to get recent recipes: %w", err)
	}

	follows, err := s.socialRepo.GetRecentFollows(ctx, targetUserID, perTypeLimit)
	if err != nil {
		return nil, fmt.Errorf("failed to get recent follows: %w", err)
	}

	reviews, err := s.socialRepo.GetRecentReviews(ctx, targetUserID, perTypeLimit)
	if err != nil {
		return nil, fmt.Errorf("failed to get recent reviews: %w", err)
	}

	favorites, err := s.socialRepo.GetRecentFavorites(ctx, targetUserID, perTypeLimit)
	if err != nil {
		return nil, fmt.Errorf("failed to get recent favorites: %w", err)
	}

	// 4. Ensure slices are not nil (return empty arrays in JSON)
	if recipes == nil {
		recipes = []dto.RecipeSummary{}
	}

	if follows == nil {
		follows = []dto.UserSummary{}
	}

	if reviews == nil {
		reviews = []dto.ReviewSummary{}
	}

	if favorites == nil {
		favorites = []dto.FavoriteSummary{}
	}

	return &dto.UserActivityResponse{
		UserID:          targetUserID.String(),
		RecentRecipes:   recipes,
		RecentFollows:   follows,
		RecentReviews:   reviews,
		RecentFavorites: favorites,
	}, nil
}

// canAccessUserActivity checks if requester can view target's activity.
func (s *SocialServiceImpl) canAccessUserActivity(
	ctx context.Context,
	requesterID *uuid.UUID,
	targetUserID uuid.UUID,
) (bool, error) {
	// User can always view their own activity
	if requesterID != nil && *requesterID == targetUserID {
		return true, nil
	}

	// Fetch privacy preferences
	privacy, err := s.userRepo.FindPrivacyPreferencesByUserID(ctx, targetUserID)
	if err != nil {
		return false, fmt.Errorf("failed to fetch privacy preferences: %w", err)
	}

	switch privacy.ProfileVisibility {
	case profileVisibilityPublic:
		return true, nil
	case profileVisibilityFollowersOnly:
		// Anonymous users cannot access followers_only profiles
		if requesterID == nil {
			return false, nil
		}
		// Check if requester follows the target user
		isFollowing, err := s.userRepo.IsFollowing(ctx, *requesterID, targetUserID)
		if err != nil {
			return false, fmt.Errorf("failed to check following status: %w", err)
		}

		return isFollowing, nil
	case profileVisibilityPrivate:
		return false, nil
	default:
		return false, nil
	}
}

func (s *SocialServiceImpl) canAccessFollowingList(
	ctx context.Context,
	requesterID, targetUserID uuid.UUID,
) (bool, error) {
	// User can always view their own following list
	if requesterID == targetUserID {
		return true, nil
	}

	// Fetch privacy preferences
	privacy, err := s.userRepo.FindPrivacyPreferencesByUserID(ctx, targetUserID)
	if err != nil {
		return false, fmt.Errorf("failed to fetch privacy preferences: %w", err)
	}

	switch privacy.ProfileVisibility {
	case profileVisibilityPublic:
		return true, nil
	case profileVisibilityFollowersOnly:
		// Check if requester follows the target user
		isFollowing, err := s.userRepo.IsFollowing(ctx, requesterID, targetUserID)
		if err != nil {
			return false, fmt.Errorf("failed to check following status: %w", err)
		}

		return isFollowing, nil
	case profileVisibilityPrivate:
		return false, nil
	default:
		return false, nil
	}
}

func (s *SocialServiceImpl) buildFollowingResponse(
	users []dto.User,
	totalCount, limit, offset int,
	countOnly bool,
) *dto.GetFollowedUsersResponse {
	response := &dto.GetFollowedUsersResponse{
		TotalCount: totalCount,
	}

	if countOnly {
		// When countOnly is true, return only totalCount (other fields remain nil)
		return response
	}

	// Ensure users is not nil (return empty slice instead)
	if users == nil {
		users = []dto.User{}
	}

	response.FollowedUsers = users
	response.Limit = &limit
	response.Offset = &offset

	return response
}
